fix: Skip over duplicated functions in SymCaches

Swatinem opened the pull request:
The `DwarfDebugSession::functions` Iterator can yield duplicated functions, which, combined with a ton of inlinees can blow out the `ParentOffset` limit.

This does fix the symptom of these duplicated functions, but I would say that the real fix would rather be somewhere in the Iterator that yields these values.

jan-auer commented:
I'll have a closer look at this PR at a later time. Just as an initial warning: Please be extremely careful with this change. IIRC, we used to have such deduplication logic and it caused issues.

Swatinem commented:
I moved the deduplication step into the `functions` iterator per our discussion and added a comment explaining what we *think* is going on.

Next steps:
* run the offending object through breakpads own `dump_syms`
* test electron before/after and see what the symcache differences are. electron stresses our symcache limits quite a bit

jan-auer commented on symbolic-symcache/src/writer.rs:
```diff
@@ -205,10 +206,15 @@ where
                 function.map_err(|e| SymCacheError::new(SymCacheErrorKind::BadDebugFile, e))?;
-            writer.add_function(function)?;
+            let fn_range = (function.address, function.size);
+            if !seen_ranges.contains(&fn_range) {
```
You can simplify this to:

```rust
if seen_ranges.insert(fn_range) {
  writer.add_function(function)?;
}
```

jan-auer commented on symbolic-symcache/src/writer.rs:
```diff
@@ -205,10 +206,15 @@ where
             .debug_session()
             .map_err(|e| SymCacheError::new(SymCacheErrorKind::BadDebugFile, e))?;
 
+        let mut seen_ranges = BTreeSet::new();
```
Would it be enough to just track the function start address, provided that we just deal with top-level functions?

jan-auer commented on symbolic-debuginfo/src/dwarf.rs:
```diff
@@ -1356,6 +1359,18 @@ impl<'s> Iterator for DwarfFunctionIterator<'s> {
+                // deduplication which merges templated code that is being generated multiple times in each
+                // compilation unit.
+                let fn_range = (func.address, func.size);
+                if self.seen_ranges.contains(&fn_range) {
```
You can simplify this to:

```rust
if !self.seen_ranges.insert((func.address, func.size)) {
  continue;
}
```

jan-auer commented on symbolic-debuginfo/src/dwarf.rs:
```diff
@@ -1356,6 +1359,18 @@ impl<'s> Iterator for DwarfFunctionIterator<'s> {
+                // functions defined in different compile units. We suspect this might be caused by link-time
+                // deduplication which merges templated code that is being generated multiple times in each
+                // compilation unit.
+                let fn_range = (func.address, func.size);
```
Would it be enough to just track the function start address, provided that we just deal with top-level functions?

jan-auer commented on symbolic-debuginfo/src/dwarf.rs:
```diff
@@ -1342,6 +1344,7 @@ impl<'s> Iterator for DwarfFileIterator<'s> {
 pub struct DwarfFunctionIterator<'s> {
     units: DwarfUnitIterator<'s>,
     functions: std::vec::IntoIter<Function<'s>>,
+    seen_ranges: BTreeSet<(u64, u64)>,
```
On second thought, MSVC may output similar duplicate entries. Can you check whether this is the case and we need to port this over, too? 

Also, I'm curious on the runtime impact on this solution. The function iterator is already quite slow due to its sorting, but we should be careful adding another _n log n_ contributor.

jan-auer commented on symbolic-debuginfo/src/dwarf.rs:
```diff
@@ -1356,6 +1359,18 @@ impl<'s> Iterator for DwarfFunctionIterator<'s> {
+                // compilation unit.
+                let fn_range = (func.address, func.size);
+                if self.seen_ranges.contains(&fn_range) {
+                    continue;
```
This check is happening too late. At this time, you have already collected the function and its full inline tree into a sorted vector. You can avoid this by pushing down `&mut seen_ranges` into `DwarfUnit::functions` a few lines below.

jan-auer commented on symbolic-debuginfo/src/dwarf.rs:
```diff
@@ -769,6 +773,18 @@ impl<'d, 'a> DwarfUnit<'d, 'a> {
+            // compilation unit. We make sure to detect this here, so we can avoid creating these
+            // duplicates as early as possible.
+            if !seen_ranges.insert((function_address, function_size)) {
+                continue;
```
Now, you need to set `skipped_depth = Some(depth);` before continuing. See other branches where we skip a function.

jan-auer commented on symbolic-debuginfo/src/dwarf.rs:
```diff
@@ -1249,6 +1267,7 @@ impl<'data> DwarfDebugSession<'data> {
             functions: Vec::new().into_iter(),
             range_buf: Vec::new(),
             finished: false,
+            seen_ranges: BTreeSet::new(),
```
_nit_: Move this before `finished`.
